src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp

```diff
@@ -204,46 +204,48 @@ WaveFunctionComponent* SlaterDetBuilder::buildComponent(xmlNodePtr cur)
       if (multislaterdet_0 || multislaterdetfast_0)
         APP_ABORT("multideterminant is already instantiated.");
 
-      std::string spo_alpha_name;
-      std::string spo_beta_name;
+      int nGroups = targetPtcl.groups();
+      std::vector<std::string> spoNames(nGroups);
+
       std::string fastAlg;
       OhmmsAttributeSet spoAttrib;
-      spoAttrib.add(spo_alpha_name, "spo_up");
-      spoAttrib.add(spo_beta_name, "spo_dn");
+      for (int grp = 0; grp < nGroups; grp++)
+        spoAttrib.add(spoNames[grp], "spo_" + std::to_string(grp));
+      if (nGroups == 2)
+      { //for legacy
+        spoAttrib.add(spoNames[0], "spo_up");
+        spoAttrib.add(spoNames[1], "spo_dn");
+      }
       spoAttrib.add(fastAlg, "Fast", {"", "yes", "no"}, TagStatus::DEPRECATED);
       spoAttrib.add(msd_algorithm, "algorithm", {"", "precomputed_table_method", "table_method", "all_determinants"});
       spoAttrib.put(cur);
 
       if (msd_algorithm.empty())
       {
-        if(fastAlg.empty())
+        if (fastAlg.empty())
           msd_algorithm = "precomputed_table_method";
-        else if(fastAlg == "yes")
+        else if (fastAlg == "yes")
           msd_algorithm = "table_method";
         else
           msd_algorithm = "all_determinants";
       }
 
-      SPOSetPtr spo_alpha = sposet_builder_factory_.getSPOSet(spo_alpha_name);
-      SPOSetPtr spo_beta  = sposet_builder_factory_.getSPOSet(spo_beta_name);
-      std::unique_ptr<SPOSet> spo_alpha_clone, spo_beta_clone;
-      if (spo_alpha == nullptr)
-      {
-        app_error() << "In SlaterDetBuilder: SPOSet \"" << spo_alpha_name
-                    << "\" is not found. Expected for MultiSlaterDeterminant.\n";
-        abort();
-      }
-      else
-        spo_alpha_clone.reset(spo_alpha->makeClone());
+      //new format
+      std::vector<SPOSetPtr> spos(nGroups);
+      std::vector<std::unique_ptr<SPOSet>> spo_clones(nGroups);
 
-      if (spo_beta == nullptr)
+      for (int grp = 0; grp < nGroups; grp++)
       {
-        app_error() << "In SlaterDetBuilder: SPOSet \"" << spo_beta_name
-                    << "\" is not found. Expected for MultiSlaterDeterminant.\n";
-        abort();
+        spos[grp] = sposet_builder_factory_.getSPOSet(spoNames[grp]);
```

Comment:
The pointer return from getSPOSet should be treated as a reference. It was kept pointer for checking nullptr. Please assign the value to a temporal pointer variable. In the current way, the SPOSet actually got destroyed when you reset the smart pointer.

Reply:
I guess I'm confused as to what you mean. The smart pointer spo_clones[grp] calls reset, but the argument is spos[grp]->makeClone(). So the spo_clones[grp] should take ownership of the pointer being returned by makeClone(), not taking ownership of spos[grp] itself.  So I don't see how the original ptr returned by getSPOSet (spos[grp]) gets destroyed.

Reply:
reset() calls delete on the previously stored pointer. This is not desired.

Reply:
Sure, but spo_clones[grp] is initially a nullptr, so it doesn't actually call the delete. I guess reset implies that it  *was* storing something before, so it is bad practice. I changed it to be a bit more clear hopefully.

Reply:
I was confused by spos and spo_clones. Could you remove spos and store the return of sposet_builder_factory_.getSPOSet in spo_tmp for nullptr check locally.
```
spo_clones.emplace_back(spo_tmp->makeClone());
```
should be OK.

Reply:
done